_posts/2021-03-05-introducing-new-window-features.md

```diff
@@ -0,0 +1,356 @@
+We only need to skip certain number of rows forward or backwards.
+
+With `RANGE`, we need to examine the actual values to see if they fall within
+the given range. Our approach is optimized for the case where the offset values
+are constant for all rows. Our solution involves caching frame bounds computed
+for the preceding row and using them as the starting point to find frame bounds
+for the current row. Ideally, we never have to move the frame bounds back as we
+process subsequent rows. In such case, the amortized cost of frame bound
+calculations per row is constant.
+
+![](/assets/blog/window-features/sliding-frame-range.svg)
+
+Our strategy for determining frame bounds for `GROUPS` is similar. We cache the
+frame bounds computed for the preceding row and use them as the starting point
+for the current row. If the frame offset is constant, frame bounds slide from
+one peer group to another every time the processed row leaves one peer group and
+enters the next one.
+
+![](/assets/blog/window-features/sliding-frame-groups.svg)
+
+### Support for WINDOW clause
+
+As all the preceding examples show, a window function is a big chunk of syntax.
+What if we wanted to use several window functions over the same window? Say, we
+need an average price and a total price from the preceding month. And the top
+price. Does it have to look like the below?
+
+```
+SELECT
+    avg(totalprice) OVER (
+        PARTITION BY custkey 
+        ORDER BY orderdate
+        RANGE BETWEEN interval '1' month PRECEDING AND CURRENT ROW),
+    sum(totalprice) OVER (
+        PARTITION BY custkey 
+        ORDER BY orderdate
+        RANGE BETWEEN interval '1' month PRECEDING AND CURRENT ROW),
+    max(totalprice) OVER (
+        PARTITION BY custkey 
+        ORDER BY orderdate
+        RANGE BETWEEN interval '1' month PRECEDING AND CURRENT ROW)
+FROM orders
+```
+
+Well, no more! In Trino, since
+[version 352]({{site.url}}/docs/current/release/release-352.html), you can
+predefine a window specification and then use it or redefine it wherever you
+need. This is thanks to the third of our new additions: support for `WINDOW`
+clause.
+
+Technically speaking, the `WINDOW` clause is part of the `FROM` clause:
+
+```
+SELECT …
+    FROM …
+        WHERE …
+        GROUP BY …
+        HAVING …
+        WINDOW …
+ORDER BY …
+OFFSET …
+LIMIT / FETCH …
+```
+
+In the `WINDOW` clause, you can define any number of named windows. Then you
+can simply refer to them by their names wherever you need them, that is, either
+in the `SELECT` list or in the `ORDER BY` clause.
+
+Let's check how the `WINDOW` clause helps with our example query:
+
+```
+SELECT 
+	avg(totalprice) OVER w,
+	sum(totalprice) OVER w,
+	max(totalprice) OVER w
+FROM orders
+WINDOW w AS (
+    PARTITION BY custkey
+    ORDER BY orderdate
+    RANGE BETWEEN interval '1' month PRECEDING AND CURRENT ROW)
+```
+
+Fon even better conciseness, the `WINDOW` clause allows you to reuse named
```

Comment:
To be even more concise, the `WINDOW` clause allows you to define more specialized windows from existing window definitions.